File: Jim/src/sk/fiit/jim/agent/highskill/GoToPosition.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.LambdaCallable;
import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.TacticalInfo;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;
import sk.fiit.robocup.library.geometry.Vector3D;

/**
 *  High skill to move player to position from parameter.
 *  Reimplemented from Ruby to Java by Martin Markech
 *  @author xsuchac
 *  @author: A55-Kickers
 *  
 */
public class GoToPosition extends HighSkill {
	private static Angles leftStrafeCrossing = new Angles(60.0, 80.0);
	private static Angles rightStrafeCrossing = new Angles(280.0, 300.0);
	private static Angles leftBackCrossing = new Angles(80.0, 100.0);
	private static Angles rightBackCrossing = new Angles(260.0, 280.0);
    private static Angles straightRange1 = new Angles(0.0, 30.0);
    private static Angles straightRange2 = new Angles(330.0, 360.0);
    private static Angles backRange = new Angles(90.0, 270.0);
    private static Angles rightRange = new Angles(180.0, 330.0);
    private static Angles leftRange = new Angles(30.0,180.0);
    private static Angles leftStrafeRange = new Angles(45.0, 90.0);
    private static Angles rightStrafeRange = new Angles(270.0, 315.0);
    private static double strafeDistance = 1.5;
    private static double veryCloseDistance = 0.4;
	
    private TacticalInfo tacticalInfo;
    private AgentInfo agentInfo;   
    private Vector3D targetPositionGlobal;
    private Vector3D targetPosition;
    private double targetPositionPhi;
    private double targetPositionPhiTemp;
    private AgentModel agentModel;
    private LambdaCallable validityProc;
    
    /*
     * @param position Vector3D Where player has to move
     * @param validityProc LambdaCallable "lambda" expression
     */
    
    public GoToPosition(Vector3D position, LambdaCallable validityProc) {
    	this.agentInfo = AgentInfo.getInstance();
    	this.agentModel = AgentModel.getInstance();
    	this.tacticalInfo = TacticalInfo.getInstance();
    	this.targetPositionGlobal = position; 
    	this.validityProc = validityProc;
    }
    
    
    
    @Override	
	public LowSkill pickLowSkill() {
    	if (!validityProc.call() ) {
    		return null;
    	}    	
    	
    	
    	if (this.targetPositionGlobal == null){
    		return null;
    	}
    	else {
    		this.agentModel = AgentModel.getInstance();
    		this.targetPosition = this.agentModel.relativize(this.targetPositionGlobal);
    	}
    	
    	 this.targetPositionPhi = this.targetPosition.getPhi();
    	
    	 if (this.targetPositionPhi != 0){
    		 this.targetPositionPhiTemp = this.targetPositionPhi;    		 
    	 }
    	 else if(this.targetPositionPhi == 0 && WorldModel.getInstance().isSeeBall()){
    		 this.targetPositionPhi = this.targetPositionPhiTemp;
    	 }
    	 
    	
    	AgentModel agentModel = AgentModel.getInstance();
    	
    	if(agentModel.isOnGround() ){
    		this.agentInfo.loguj("on ground ");
    		return null;
    	}
    	else if (isOnPosition()){
    		this.agentInfo.loguj("close ");
    		return null;
    	}
    	else if (isLeftAndClose()){
    		return LowSkills.get("step_left");
    	}
    	else if(isRightAndClose()){
    		return LowSkills.get("step_right");
    	}
    	else if(isBackAndClose() ){
    		return LowSkills.get("walk_back");
    	}
    	else if(isStraight() ){
    		return LowSkills.get("walk_forward");
    	}
    	else if(isLeftAndDistant() ){
    		return LowSkills.get("turn_left_cont_20");
    	}
    	else if(isRightAndDistant() ){
    		return LowSkills.get("turn_right_cont_20");
    	}
    	
    	
    	return null;
	}
    
    private boolean isCloseEnough(){
    	return this.targetPosition.getR() < 0.5;
    }
    
    private boolean isStraight(){
    	return this.straightRange1.include(this.targetPositionPhi) || this.straightRange2.include(this.targetPositionPhi);
    }
    
    private boolean isBackAndClose(){
    	return this.backRange.include(this.targetPositionPhi) && (this.targetPosition.getR() <= this.strafeDistance);
    	  
    }
    
    private boolean isLeftAndClose(){
    	return (this.targetPosition.getR() <= this.strafeDistance) && this.leftStrafeRange.include(Angles.normalize(this.targetPositionPhi));
    }
   
    
    private boolean isRightAndClose(){
    	return (this.targetPosition.getR() <= this.strafeDistance) && this.rightStrafeRange.include(Angles.normalize(this.targetPositionPhi)); 
    }
    
    private boolean isLeftAndDistant(){
    	return this.leftRange.include(Angles.normalize(this.targetPositionPhi));
    }
    
    private boolean isRightAndDistant(){
    	return this.rightRange.include(Angles.normalize(this.targetPositionPhi));
    }
   

    //Todo #Task(During rewriting from Ruby - seems it is not used or brokend) #Solver() #Priority(tiny) | xmarkech 2013-12-10T20:36:57.5950000Z
    /*   
    private boolean isBack(double target_position){
    	return Angles.angleDiff(target_phi, Math.PI) < Math.PI / 6.0;    	
    }
*/    
    
    
    private boolean isOnPosition(){
    	return Math.abs(this.agentModel.getPosition().getXYDistanceFrom(this.targetPositionGlobal)) < 0.5;
    }
    
    //Todo #Task(Check if it is needed after migrating to new architecture) #Solver() #Priority() | xmarkech 2013-12-10T20:38:13.4560000Z 
	@Override
	public void checkProgress() throws Exception {
		// TODO Auto-generated method stub
		
	}
    
	
}

File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/GoToPosition.rb
##### goToPosition.rb ######
#	High skill to move player to position from parameter.
#	@author: xsuchac
#	@author: A55-Kickers
#	@param position - where player has to move,
#		type sk.fiit.robocup.library.geometry.Vector3D
#	@param validity_proc - process for calling method from plan to check
#		if GoToPosition is still valid or not
#
#	Example of instatiating with parameter of coordinates [x,y,z]:
#	GoToPosition.new(Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(x, y, z), Proc.new{see_ball?})
############################

require 'high_skills/ruby_high_skill.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math

class GoToPosition < RubyHighSkill

	def GoToPosition.angle_range from, to
    	(from/180.0*Math::PI)..(to/180.0*Math::PI)
  	end

  #65 - 85 - the same as above
  @@left_strafe_crossing = angle_range(60.0, 80.0)
  #275..295
  @@right_strafe_crossing = angle_range(280.0, 300.0)
  ##same...
  @@left_back_crossing = angle_range(80.0, 100.0)
  @@right_back_crossing = angle_range(260.0, 280.0)

  #careful - do not put 330 instead of -30, will cause havoc in Ruby range, as expected (it doesn't know it is dealing with degrees)
  @@straight_range1 = angle_range(0.0, 30.0)
  @@straight_range2 = angle_range(330.0, 360.0)
  @@back_range = angle_range(90.0, 270.0)
  #270 - 330 degrees
  @@right_range = angle_range(180.0, 330.0)
  #30 - 90 degrees
  @@left_range = angle_range(30.0, 180.0)
  #75 - 90 degrees
  @@left_strafe_range = angle_range(45.0, 90.0)
  #270 - 285 degrees
  @@right_strafe_range = angle_range(270.0, 315.0)
  @@strafe_distance = 1.5
  @@very_close_distance = 0.4
  
  def initialize position, validity_proc
    super()
    @agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
    @agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
    @tacticalInfo = Java::sk.fiit.jim.agent.models.TacticalInfo.getInstance
    
    @target_position_global = position
    @validity_proc = validity_proc

  end
  
  def pickLowSkill
	 if not @validity_proc.call
	 	return nil
	 end
  
  	 if @target_position_global == nil
  	 	return nil
  	 else
  	 	@agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
  	 	@target_position = @agentModel.relativize(@target_position_global)
  	 end
    
     @target_position_phi = @target_position.getPhi()
     
     if @target_position_phi != nil
     	@target_position_phi2 = @target_position_phi
     elsif @target_position_phi == nil and plan.see_ball?
     	@target_position_phi = @target_position_phi2
     end
  
    case
		when me.on_ground?
			#puts "--onGround"
      		return	
       	when is_on_position?
       		@agentInfo.loguj("close ")
       		#puts "--onPosition"
  			return nil
        when left_and_close?:
        	#puts "--leftAndClose"
        	return get_skill("step_left")
      	when right_and_close?:
      		#puts "--rightAndClose"
        	return get_skill("step_right")
      	when back_and_close?:
      		#puts "--backAndClose"
        	return get_skill("walk_back")
        when straight?:
      		#puts "--straight"
        	return get_skill("walk_forward")
     	when left_and_distant?:
     		#puts "--turnLeft"
        	return get_skill("turn_left_cont_20")
     	when right_and_distant?:
     		#puts "--turnRight"
        	return get_skill("turn_right_cont_20")
        else
        	return nil
        
    end   
  end
  
  def is_current_skill? possibilities
    current_skill_name = currentSkill.nil? ? "null" : currentSkill.name
    possibilities.member?(current_skill_name)
  end
  
  def close_enough?
    @target_position.getR < 0.5
  end
  
  def is_on_position?
  	abs(@agentModel.getPosition().getXYDistanceFrom(@target_position_global)) < 0.5
  end

  def straight?
    @@straight_range1.include?(@target_position_phi) or @@straight_range2.include?(@target_position_phi)
  end
  
  def back_and_close?
    @@back_range.include?(@target_position_phi) and (@target_position.getR <= @@strafe_distance)
  end

  def left_and_close?
    (@target_position.r <= @@strafe_distance) and @@left_strafe_range.include?(Angles.normalize(@target_position_phi))
  end

  def right_and_close?
    (@target_position.r <= @@strafe_distance) and @@right_strafe_range.include?(Angles.normalize(@target_position_phi))
  end

  def left_and_distant?
    @@left_range.include? Angles.normalize(@target_position_phi)
  end

  def right_and_distant?
    @@right_range.include? Angles.normalize(@target_position.getPhi())
  end
  
  def back? target_position
    Angles.angle_diff(target_phi, Math::PI) < Math::PI / 6.0
  end
  
  def back? target_position
    Angles.angle_diff(target_phi, Math::PI) < Math::PI / 6.0
  end
  
  def checkProgress
#    raise "Fallen!" if me.on_ground?
  end
  
  def abs number
    Math.abs number
  end
end

#in order not to defer the startup cost
GoToPosition.new nil, nil
